### DISCUSSION TITLE
CSV to SLP, and further training

### DISCUSSION THREAD

**Discussion by jramborger78** [talmolab/sleap#1887] [Help!]
Hey guys,

Hope you are all well. Right now I am trying to import my SLEAP (.slp) interpolated results from a tracking of a video into SimBA. Speaking with their discussion board, they adapted a patch for us as we use 1 animal in a chamber as opposed to multi-animals, which was the prior method since their paper release.

That being said, I am trying to only import the columns that contain the rat coordinates, not the other 6 columns (9 with scores) of the active and inactive levers, and cue light. I keep running into issues trying to do so myself, not super familiar with H5 formats and what not, so was curious if there was a way to:

1. convert CSV to slp
2. given I am attempting to remove the last 9 columns, a quick code to convert the slp to H5 and for it to remove those columns.

Also, recently the model we have worked wonderfully for one of our cohorts of rats. Just finished analyzing another and noticed some tracking did pretty poorly. My best guess is from cohort to cohort the same chamber will have a different color or sized rat amongst other features, as well as any lighting changes form the box moving around in the SAC, etc.. My intuition tells me that, if i take a couple videos from the new cohort, maybe those specifically that are doing poorly, label some frames, and then continue training the model with the new additions would be the best approach to improve the model as opposed to completely train another. So I guess my question is:

1. is my intuition correct here
2. Is the best method to then "resume" training of the model selected, or to "use trained model", or simply, train a new one with the starting weights being the prior model.

Thanks as always guys, George Lab appreciates you!
-Jarryd

**Comment by roomrys** [comment]
Hi @jramborger78,

For the CSV conversion, we would need to know what the format of the CSV is. I wrote a [gist that converts a CSV](https://gist.github.com/roomrys/4f86de88368848afe15a315264cdfca3) of a certain structure to an SLP. Not sure how comfortable you feel modifying the code - otherwise, you can provide us with an example CSV (we really just need the headers and a few example rows).

The code provided in the gist expects the following headers/example data:
| track | frame_idx | instance.score | Nose.x | Nose.y | Nose.score | Head.x | Head.y | Head.score | Neck.x | Neck.y | Neck.score | Back.x | Back.y | Back.score | TailBase.x | TailBase.y | TailBase.score |
|-------|-----------|----------------|--------|--------|------------|--------|--------|------------|-------------|-------------|-----------------|-------------|-------------|-----------------|------------|------------|----------------|
| many  | 0         |                |        | 756.481784 | 3.112872624 |        | 744.0468087 | 2.98986738 |             | 720.8135945 | 7.80031706 |             | 708.3836689 | 12.89300192 |            |            |                |

and recreates and recreates an SLP file from the data.

Your intuition is correct - that is what we would suggest to generalize to new videos. You should either use "resume" or start training from scratch. The "resume" option comes with the con that you may get stuck in a local minima that the network was previously trained to. The pro would be that you could reach your new minima faster.

Thanks,
Liezl

**Comment by jramborger78** [comment]
Appreciate that, thanks! Good points. For csvs, basically i remove the scores columns for each xy and the track column, but simba appears to expect those. So when all is said and done it is a frames column and the xy for each node. Sincerely,Jarryd Ramborger

**Reply by roomrys** [reply]
Hi @jramborger78,

Does the gist I provided work for you? 

The [csv to slp conversion gist](https://gist.github.com/roomrys/4f86de88368848afe15a315264cdfca3) expects the following headers:
| track | frame_idx | instance.score | node_name.x | node_name.y | node_name.score |
|-------|-----------|----------------|--------|--------|------------|
| track_1| 0         |                |        | 756.481784 | 3.112872624 |        |

where the `node_name` headers are replaced with the actual `Node`/body-part names (there can be multiple of these).

If you do not want to add a `track`, then the `track` column ~~should exist, but~~ **can either be removed or** be left/set-to empty. If all the `node_name.score`s are empty for a row, then that annotation will be created as a user-labeled `Instance`. Otherwise, the annotation for that row is created as a `PredictedInstance`.

To run the gist, you would need to download the file, make changes to the file paths in the `__main__` block:
```python
if __name__ == "__main__":
    # Specify the file path
    file_path = r"c:\path\to.csv"  # Path to CSV file with SLEAP data
    video_filename = (  # Path to video used for annotations (assuming one video per CSV)
        r"c:\path\to\video.avi"
    )
    output_filename = r"test.slp"  # Path to save output SLP file to
```

then activate your sleap environment and run the python file with:
```bash
python sleap_csv_to_slp.py
```

To skip over specific nodes/scores, you can also add them to the list of `non_node_headers` in the `__main__` block:
```python
if __name__ == "__main__":
    ...  # Some other stuff shown above
    non_node_headers = ["frame_idx", "track", "instance.score", "node_to_exclude.x"]
```
where `node_to_exclude` should be replaced with the `Node`/body-part name to exclude. The code only looks for the `.x` header, retrieves the node name from the header, and uses the node name to find the `.y` and `.score` data. The `.score` data column does not need to exist, but `.x` and `.y` do.

---

If this does not work for you, do you mind providing me with the exact header names, an example CSV, or the error message when you try running the gist.

Thanks,
Liezl

**Reply by jramborger78** [reply]
Hi Liezl,

Thank you as always for your assistance. May be easier to explain as such.

So when I track in sleap on videos, i export as H5 to run the interpolation code you have in the analysis examples page, I then convert it back to an slp file (so that I have an interpolated version to check tracking), and then have sleap export a csv of the results, as you would in the GUI. That csv i then read to then clean aways the track and scores columns, and add the subject's name to the column names.

In SimBA, their conversion code to import pose-estimations expects those columns I remove. So turning the csv to slp may not work given I clean those away and change the column names by adding the subject name.

So instead what I am after is a way to convert the slp to H5, and then remove the last 3 nodes. SimBA loads in H5 files faster than SLP files, so it can stay as an H5. 

I would then use that code snippet to add to my current track code that spits out the usual csvs for statistics, but also an H5 file that is "cleaned" for SimBA with those 3 nodes removed. Best I have intuited it operating in the same directory is like this:

```python

# Proxy examples
main_path = "C:\\Users\\jramb\\Documents"
video_name_no_ext = "C25LGA01_Box02_2024-02-05"


# Step 1: Interpolate
h5_file_path = f"{main_path}/{session_name}_{subject}.h5"
with h5py.File(h5_file_path, "a") as f:
    # Transpose the data
    locations = f["tracks"][:].T
    # Interpolate
    locations = fill_missing(locations)
    # Transpose back
    f["tracks"][:] = locations.T


print(f"interpolated data saved as {session_name}_{subject}.h5")


# Step 2: Rename a copy for SimBA
new_h5_file_path = f"{main_path}/{video_name_no_ext}.h5"
shutil.copy(h5_file_path, new_h5_file_path)


# Step 3: Remove last 3 nodes for SimBA
with h5py.File(new_h5_file_path, "a") as f:
    locations = f["tracks"][:].T
    # Slice the array to remove the last 3 nodes
    locations = locations[:, :-3, :, :]
    # Transpose back
    f["tracks"][:] = locations.T

print(f"Nodes removed and saved as {video_name_no_ext}.h5")

```

**Reply by jramborger78** [reply]
Following up, i figured out how to remove the nodes. Thank you for the help!

Also, wanted to show the model progress.

I resumed training on the old model on the new videos till i saw good numbers with 600 labeled frames, merged it with the old labels file of 1350 frames, added 50 more for 2000 total, and did a few rounds of resumed training. Also, inference on a video that was not in the training videos that tracked poorly with the old model ended up having better looking predictions, especially on the catheter line hat was bouncing from the point on the back (where it is supposed to predict) to up the coil a bit often at times. Seems to have worked!

![image](https://github.com/user-attachments/assets/18350b21-8980-4c4d-9314-fc9f585538e1)

**Comment by jramborger78** [comment]
Something that may at least move forward in the meantime with present files is something that allows me to export an slp to H5 but removes what would be the "last 9 columns" in a csv, to trim the levers and cue xy's and score columns. I could then take that and put it into my tracking code that runs on a super pcc to spit out my usual files as well as an H5 that is doctored for SimBA.
